server/src/instant/db/cel.clj

```diff
@@ -265,8 +337,621 @@
+                     (assoc acc k (cond (map? v)
+                                        (cond (contains? v :$not)
+                                              (:$not v)
+
+                                              (contains? v :$isNull)
+                                              {:$isNull (not (:$isNull v))}
+
+                                              :else (throw (ex-info "Unsupported operation for negation", {:v v})))
+
+                                        :else
+                                        {:$not v})))
+                   {}
+                   c)))
+
+(def not-overloads
+  (global-overload
+   (.getFunction Operator/LOGICAL_NOT)
+   [{:overload-id "_not_datakey"
+     :cel-args [datakey-cel-type]
+     :cel-return-type whereclause-cel-type
+     :java-args [DataKey]
+     :impl (fn [[^DataKey x]]
+             (WhereClause. {(.data_key x) false}))}
+    {:overload-id "_not_whereclause"
+     :cel-args [whereclause-cel-type]
+     :cel-return-type whereclause-cel-type
+     :java-args [WhereClause]
+     :impl (fn [[^WhereClause x]]
+             (WhereClause. (negate-where-clauses (.where_clause x))))}]))
+
+;; Overloads for `startsWith`, `endsWith`, and `contains`
+
+(def starts-with-overload
+  (member-overload
+   "startsWith"
+   [{:overload-id "_datakey_starts_with"
+     :cel-args [datakey-cel-type SimpleType/STRING]
+     :cel-return-type whereclause-cel-type
+     :java-args [DataKey String]
+     :impl (fn [[^DataKey x ^String y]]
+             (let [attr (.instant_attr x)]
+               (if (and (= :string (:checked-data-type attr))
+                        (:index? attr))
+                 (WhereClause. {(.data_key x) {:$like (str y "%")}})
+                 (throw (ex-info "Invalid attr" {:x x})))))}]))
+
+(def ends-with-overload
+  (member-overload
+   "endsWith"
+   [{:overload-id "_datakey_ends_with"
+     :cel-args [datakey-cel-type SimpleType/STRING]
+     :cel-return-type whereclause-cel-type
+     :java-args [DataKey String]
+     :impl (fn [[^DataKey x ^String y]]
+             (let [attr (.instant_attr x)]
+               (if (and (= :string (:checked-data-type attr))
+                        (:index? attr))
+                 (WhereClause. {(.data_key x) {:$like (str "%" y)}})
+                 (throw (ex-info "Invalid attr" {:x x})))))}]))
+
+(def contains-overload
+  (member-overload
+   "contains"
+   [{:overload-id "_datakey_contains"
+     :cel-args [datakey-cel-type SimpleType/STRING]
+     :cel-return-type whereclause-cel-type
+     :java-args [DataKey String]
+     :impl (fn [[^DataKey x ^String y]]
+             (let [attr (.instant_attr x)]
+               (if (and (= :string (:checked-data-type attr))
+                        (:index? attr))
+                 (WhereClause. {(.data_key x) {:$like (str "%" y "%s")}})
+                 (throw (ex-info "Invalid attr" {:x x})))))}]))
+
+;; Overloads for `type` to prevent `type(data.key)` from succeeding
```

Comment:
Realized, in the future I guess we could have these return _actual_ underlying types! (i.e if data.title is a string etc)